flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DeclarativeSlotPoolBridge.java

```diff
@@ -317,11 +317,12 @@ private PhysicalSlot reserveFreeSlotForResource(SlotRequestId slotRequestId, All
 					timeout.toMilliseconds(),
 					TimeUnit.MILLISECONDS,
 					componentMainThreadExecutor)
-				.whenComplete((physicalSlot, throwable) -> {
-					if (throwable instanceof TimeoutException) {
-						timeoutPendingSlotRequest(slotRequestId);
-					}
-				});
+				.whenCompleteAsync((physicalSlot, throwable) -> {
+						if (throwable instanceof TimeoutException) {
+							timeoutPendingSlotRequest(slotRequestId);
+						}
+					},
+					componentMainThreadExecutor);
```

Comment:
After an offline discussion with Chesnay we concluded that this is not the fix for the problem. Instead we will try to not expose futures which have other dependents because this can lead to violating the main thread guarantee when waiting on this future outside of the main thread (`get()` will try to complete the future's dependents).

Reply:
Due to `whenCompleteAsync` the `testRequirementsDecreasedOnAllocationTimeout` now sometimes fails because the `timeoutPendingSlotRequest` is not run before the final assertion.